File: tests/Tests.FSharp/CloudEvents.Tests.fs
module Zeta.Tests.CloudEventsTests

open global.Xunit
open Zeta.Core
open System
open System.IO
open System.Text.Json
open System.Globalization
open System.Collections.Immutable

module CE = Zeta.Core.CloudEvents

type private Marker = class end

let private repoRoot () : string =
    let assembly = typeof<Marker>.Assembly
    let mutable dir = DirectoryInfo(Path.GetDirectoryName(assembly.Location))
    while not (isNull dir) && not (File.Exists(Path.Join(dir.FullName, "Zeta.sln"))) do
        dir <- dir.Parent
    if isNull dir then
        raise (InvalidOperationException("Could not locate repo root (Zeta.sln) from test assembly location."))
    dir.FullName

let rec private buildValue (el: JsonElement) : DynamicValue =
    let tag = el.GetProperty("t").GetString()
    match tag with
    | "null" -> DynamicValue.Null
    | "bool" -> DynamicValue.Bool(el.GetProperty("v").GetBoolean())
    | "int" -> DynamicValue.Int(int64 (el.GetProperty("v").GetString()))
    | "float" ->
        let bits = UInt64.Parse(el.GetProperty("v").GetString(), NumberStyles.HexNumber, CultureInfo.InvariantCulture)
        DynamicValue.Float(BitConverter.UInt64BitsToDouble(bits))
    | "str" -> DynamicValue.String(el.GetProperty("v").GetString())
    | "bytes" ->
        let hexStr = el.GetProperty("v").GetString()
        DynamicValue.Bytes(ImmutableArray.Create<byte>(Convert.FromHexString(hexStr)))
    | "arr" ->
        let arr = el.GetProperty("v").EnumerateArray() |> Seq.map buildValue |> Seq.toList
        DynamicValue.Array(arr)
    | "obj" ->
        let obj =
            el.GetProperty("v").EnumerateArray()
            |> Seq.map (fun pair ->
                let parts = pair.EnumerateArray() |> Seq.toArray
                parts.[0].GetString(), buildValue parts.[1]
            )
            |> Seq.toList
        DynamicValue.Object(obj)
    | _ -> failwithf "unsupported tag: %s" tag

let private tryStr (el: JsonElement) (prop: string) : string option =
    match el.TryGetProperty(prop) with
    | true, p when p.ValueKind = JsonValueKind.String -> Some(p.GetString())
    | _ -> None

[<Fact>]
let ``create yields a valid v1.0 event; validate catches a missing required attribute`` () =
    let e = CE.create "id-1" "/zeta/source" "com.zeta.change" (Some(DynamicValue.Int 7L))
    Assert.Equal("1.0", e.SpecVersion)
    Assert.Equal<Result<unit, string>>(Ok(), CE.validate e)
    match CE.validate { e with Id = "" } with
    | Error msg -> Assert.Contains("id", msg)
    | Ok () -> Assert.Fail "expected missing-id error"

[<Fact>]
let ``toDynamic ∘ ofDynamic round-trips (required + optionals + extensions + data)`` () =
    let e =
        { CE.create "id-2" "/s" "t" (Some(DynamicValue.String "payload")) with
            Time = Some "2026-06-07T00:00:00Z"
            DataSchema = Some "schema://v2"
            Extensions = [ "iodebeziumop", "c"; "traceparent", "abc" ] }
    Assert.Equal<Result<CE.CloudEvent, string>>(Ok e, CE.ofDynamic (CE.toDynamic e))

[<Fact>]
let ``ofDynamic rejects a non-object and an object missing required attributes`` () =
    Assert.True(match CE.ofDynamic (DynamicValue.Int 1L) with Error _ -> true | _ -> false)
    Assert.True(
        match CE.ofDynamic (DynamicValue.Object [ "id", DynamicValue.String "x" ]) with
        | Error _ -> true
        | _ -> false
    ) // missing source/type

[<Fact>]
let ``unknown string keys become extension attributes, core keys do not`` () =
    let dv =
        DynamicValue.Object
            [ "specversion", DynamicValue.String "1.0"
              "id", DynamicValue.String "i"
              "source", DynamicValue.String "s"
              "type", DynamicValue.String "t"
              "myext", DynamicValue.String "v"
              "data", DynamicValue.Int 5L ]
    match CE.ofDynamic dv with
    | Ok e ->
        Assert.Equal<(string * string) list>([ "myext", "v" ], e.Extensions)
        Assert.Equal<DynamicValue option>(Some(DynamicValue.Int 5L), e.Data)
    | Error m -> Assert.Fail m

[<Fact>]
let ``CrossVerifyCloudEventVectorsMatchExpected`` () =
    let root = repoRoot ()
    let jsonPath = Path.Combine(root, "tests", "cross-verification", "dv-key-cloud-events", "vectors.json")
    Assert.True(File.Exists(jsonPath), sprintf "vectors.json not found: %s" jsonPath)

    use doc = JsonDocument.Parse(File.ReadAllText(jsonPath))
    let vectors = doc.RootElement.GetProperty("cloud_event_vectors").EnumerateArray()

    for v in vectors do
        let expectedJson = v.GetProperty("expected_json").GetString()
        let expectedCborHex = v.GetProperty("expected_cbor_hex").GetString()

        let eventEl = v.GetProperty("event")
        let id = eventEl.GetProperty("id").GetString()
        let source = eventEl.GetProperty("source").GetString()
        let typ = eventEl.GetProperty("type").GetString()
        let specversion = eventEl.GetProperty("specversion").GetString()

        let data =
            match eventEl.TryGetProperty("data") with
            | true, d when d.ValueKind <> JsonValueKind.Null -> Some(buildValue d)
            | _ -> None

        let extensions =
            match eventEl.TryGetProperty("extensions") with
            | true, ext when ext.ValueKind = JsonValueKind.Array ->
                ext.EnumerateArray()
                |> Seq.map (fun pair ->
                    let parts = pair.EnumerateArray() |> Seq.toArray
                    parts.[0].GetString(), parts.[1].GetString()
                )
                |> Seq.toList
            | _ -> []

        let ce =
            { CE.create id source typ data with
                SpecVersion = specversion
                Time = tryStr eventEl "time"
                Subject = tryStr eventEl "subject"
                DataContentType = tryStr eventEl "datacontenttype"
                DataSchema = tryStr eventEl "dataschema"
                Extensions = extensions }

        let dynamicVal = CE.toDynamic ce

        // JSON Canonical check
        let actualJson =
            match DynamicValue.toCanonicalJson dynamicVal with
            | Ok s -> s
            | Error e -> failwithf "JSON encode failed: %A" e
        Assert.Equal(expectedJson, actualJson)

        // CBOR Canonical check
        let actualCborHex =
            match DynamicValue.toCanonicalCbor dynamicVal with
            | Ok bytes -> Convert.ToHexString(bytes).ToLowerInvariant()
            | Error e -> failwithf "CBOR encode failed: %A" e
        Assert.Equal(expectedCborHex, actualCborHex)

File: tests/Tests.CSharp/CloudEventsTests.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text.Json;
using Xunit;
using Zeta.Core.CSharp;

namespace Zeta.Tests.CSharp;

public class CloudEventsTests
{
    private static string RepoRoot()
    {
        var dir = new DirectoryInfo(Path.GetDirectoryName(typeof(CloudEventsTests).Assembly.Location)!);
        while (dir is not null && !File.Exists(Path.Join(dir.FullName, "Zeta.sln")))
        {
            dir = dir.Parent;
        }
        return dir?.FullName ?? throw new DirectoryNotFoundException("Could not locate repo root (Zeta.sln).");
    }

    private static string Str(JsonElement el, string prop) =>
        el.GetProperty(prop).GetString()
            ?? throw new InvalidOperationException($"fixture property '{prop}' is not a string");

    private static string? TryStr(JsonElement el, string prop)
    {
        if (el.TryGetProperty(prop, out var p) && p.ValueKind == JsonValueKind.String)
        {
            return p.GetString();
        }
        return null;
    }

    private static DynamicValue BuildValue(JsonElement el)
    {
        string tag = Str(el, "t");
        switch (tag)
        {
            case "null":
                return new DynamicValue.Null();
            case "bool":
                return new DynamicValue.Bool(el.GetProperty("v").GetBoolean());
            case "int":
                return new DynamicValue.Int(long.Parse(Str(el, "v"), CultureInfo.InvariantCulture));
            case "float":
                return new DynamicValue.Float(
                    BitConverter.UInt64BitsToDouble(
                        ulong.Parse(Str(el, "v"), NumberStyles.HexNumber, CultureInfo.InvariantCulture)));
            case "str":
                return new DynamicValue.String(Str(el, "v"));
            case "bytes":
                return new DynamicValue.Bytes(Convert.FromHexString(Str(el, "v")).ToImmutableArray());
            case "arr":
                return new DynamicValue.Array(
                    el.GetProperty("v").EnumerateArray().Select(BuildValue).ToImmutableArray());
            case "obj":
                return new DynamicValue.Object(
                    el.GetProperty("v").EnumerateArray()
                        .Select(pair =>
                        {
                            JsonElement[] parts = pair.EnumerateArray().ToArray();
                            string key = parts[0].GetString() ?? "";
                            return new KeyValuePair<string, DynamicValue>(key, BuildValue(parts[1]));
                        })
                        .ToImmutableArray());
            default:
                throw new InvalidOperationException($"unsupported tag: {tag}");
        }
    }

    [Fact]
    public void CreateYieldsAValidEventAndValidateCatchesAMissingRequiredAttribute()
    {
        var e = CloudEvents.Create("id-1", "/zeta/source", "com.zeta.change", new DynamicValue.Int(7L));
        Assert.Equal("1.0", e.SpecVersion);
        Assert.True(CloudEvents.Validate(e) is Result<bool, string>.Ok);

        var missingId = e with { Id = string.Empty };
        var validationResult = CloudEvents.Validate(missingId);
        Assert.True(validationResult is Result<bool, string>.Err);
        var err = (Result<bool, string>.Err)validationResult;
        Assert.Contains("id", err.Error, StringComparison.Ordinal);
    }

    [Fact]
    public void ToDynamicOfDynamicRoundTrips()
    {
        var extensions = ImmutableArray.Create(
            new KeyValuePair<string, string>("iodebeziumop", "c"),
            new KeyValuePair<string, string>("traceparent", "abc")
        );

        var e = CloudEvents.Create("id-2", "/s", "t", new DynamicValue.String("payload")) with
        {
            Time = "2026-06-07T00:00:00Z",
            DataSchema = "schema://v2",
            Extensions = extensions
        };

        var dynamicVal = CloudEvents.ToDynamic(e);
        var ofDynamicResult = CloudEvents.OfDynamic(dynamicVal);
        Assert.True(ofDynamicResult is Result<CloudEvent, string>.Ok);
        var ok = (Result<CloudEvent, string>.Ok)ofDynamicResult;
        Assert.Equal(e, ok.Value);
    }

    [Fact]
    public void OfDynamicRejectsNonObjectAndObjectMissingRequiredAttributes()
    {
        Assert.True(CloudEvents.OfDynamic(new DynamicValue.Int(1L)) is Result<CloudEvent, string>.Err);

        var pairs = ImmutableArray.Create(new KeyValuePair<string, DynamicValue>("id", new DynamicValue.String("x")));
        var missingAttrs = new DynamicValue.Object(pairs);
        Assert.True(CloudEvents.OfDynamic(missingAttrs) is Result<CloudEvent, string>.Err);
    }

    [Fact]
    public void UnknownStringKeysBecomeExtensionAttributesAndCoreKeysDoNot()
    {
        var pairs = ImmutableArray.Create(
            new KeyValuePair<string, DynamicValue>("specversion", new DynamicValue.String("1.0")),
            new KeyValuePair<string, DynamicValue>("id", new DynamicValue.String("i")),
            new KeyValuePair<string, DynamicValue>("source", new DynamicValue.String("s")),
            new KeyValuePair<string, DynamicValue>("type", new DynamicValue.String("t")),
            new KeyValuePair<string, DynamicValue>("myext", new DynamicValue.String("v")),
            new KeyValuePair<string, DynamicValue>("data", new DynamicValue.Int(5L))
        );

        var dv = new DynamicValue.Object(pairs);
        var ofDynamicResult = CloudEvents.OfDynamic(dv);
        Assert.True(ofDynamicResult is Result<CloudEvent, string>.Ok);
        var ok = (Result<CloudEvent, string>.Ok)ofDynamicResult;

        Assert.Single(ok.Value.Extensions);
        Assert.Equal("myext", ok.Value.Extensions[0].Key);
        Assert.Equal("v", ok.Value.Extensions[0].Value);
        Assert.Equal(new DynamicValue.Int(5L), ok.Value.Data);
    }

    [Fact]
    public void CrossVerifyCloudEventVectorsMatchExpected()
    {
        var root = RepoRoot();
        var jsonPath = Path.Combine(root, "tests", "cross-verification", "dv-key-cloud-events", "vectors.json");
        Assert.True(File.Exists(jsonPath), $"vectors.json not found: {jsonPath}");

        using var doc = JsonDocument.Parse(File.ReadAllText(jsonPath));
        var vectors = doc.RootElement.GetProperty("cloud_event_vectors").EnumerateArray();

        foreach (var v in vectors)
        {
            var expectedJson = Str(v, "expected_json");
            var expectedCborHex = Str(v, "expected_cbor_hex");

            var evtEl = v.GetProperty("event");
            var evtId = Str(evtEl, "id");
            var evtSource = Str(evtEl, "source");
            var evtType = Str(evtEl, "type");
            var specversion = Str(evtEl, "specversion");

            DynamicValue? dataVal = null;
            if (evtEl.TryGetProperty("data", out var dEl) && dEl.ValueKind != JsonValueKind.Null)
            {
                dataVal = BuildValue(dEl);
            }

            var extensionsBuilder = ImmutableArray.CreateBuilder<KeyValuePair<string, string>>();
            if (evtEl.TryGetProperty("extensions", out var extEl))
            {
                foreach (var pair in extEl.EnumerateArray())
                {
                    JsonElement[] parts = pair.EnumerateArray().ToArray();
                    string extKey = parts[0].GetString() ?? "";
                    string extVal = parts[1].GetString() ?? "";
                    extensionsBuilder.Add(new KeyValuePair<string, string>(extKey, extVal));
                }
            }

            var ce = CloudEvents.Create(evtId, evtSource, evtType, dataVal) with
            {
                SpecVersion = specversion,
                Time = TryStr(evtEl, "time"),
                Subject = TryStr(evtEl, "subject"),
                DataContentType = TryStr(evtEl, "datacontenttype"),
                DataSchema = TryStr(evtEl, "dataschema"),
                Extensions = extensionsBuilder.ToImmutable()
            };

            var dynamicVal = CloudEvents.ToDynamic(ce);

            var jsonResult = DynamicValues.ToCanonicalJson(dynamicVal);
            Assert.True(jsonResult is Result<string, EncodeError>.Ok);
            var actualJson = ((Result<string, EncodeError>.Ok)jsonResult).Value;
            Assert.Equal(expectedJson, actualJson);

            var cborResult = DynamicValues.ToCanonicalCbor(dynamicVal);
            Assert.True(cborResult is Result<byte[], EncodeError>.Ok);
            var actualCborHex = Convert.ToHexString(((Result<byte[], EncodeError>.Ok)cborResult).Value).ToLowerInvariant();
            Assert.Equal(expectedCborHex, actualCborHex);
        }
    }
}
